bbot/modules/internal/excavate.py

```diff
@@ -1,476 +1,858 @@
-            event_data = {"host": str(host), "description": f"Non-HTTP URI: {result}"}
-            parsed_url = getattr(event, "parsed_url", None)
-            if parsed_url:
-                event_data["url"] = parsed_url.geturl()
-            await self.excavate.emit_event(
-                event_data,
-                "FINDING",
-                parent=event,
-                abort_if=abort_if,
-                context=f"{{module}}'s URL extractor found {{event.type}}: Non-HTTP URI {{event.data}} in {context}",
-            )
-            protocol_data = {"protocol": parsed_uri.scheme, "host": str(host)}
-            if port:
-                protocol_data["port"] = port
-            await self.excavate.emit_event(
-                protocol_data,
-                "PROTOCOL",
-                parent=event,
-                abort_if=abort_if,
-                context=f"{{module}}'s URL extractor found {{event.type}}: {{event.data}} in {result} while searching {context}",
-            )
-            return
+    options = {
+        "retain_querystring": False,
+        "yara_max_match_data": 2000,
+        "custom_yara_rules": "",
+    }
+    options_desc = {
+        "retain_querystring": "Keep the querystring intact on emitted WEB_PARAMETERS",
+        "yara_max_match_data": "Sets the maximum amount of text that can extracted from a YARA regex",
+        "custom_yara_rules": "Include custom Yara rules",
+    }
+    scope_distance_modifier = None
 
-        return self.excavate.make_event(
-            result,
-            "URL_UNVERIFIED",
-            parent=event,
-            context=f"{{module}}'s URL extractor ({name} regex) found {{event.type}}: {{event.data}} in {context}",
+    _max_event_handlers = 8
+
+    parameter_blacklist = [
+        "__VIEWSTATE",
+        "__EVENTARGUMENT",
+        "__EVENTVALIDATION",
+        "__EVENTTARGET",
+        "__EVENTARGUMENT",
+        "__VIEWSTATEGENERATOR",
+        "__SCROLLPOSITIONY",
+        "__SCROLLPOSITIONX",
+        "ASP.NET_SessionId",
+        "JSESSIONID",
+        "PHPSESSID",
+    ]
+
+    yara_rule_name_regex = re.compile(r"rule\s(\w+)\s{")
+    yara_rule_regex = re.compile(r"(?s)((?:rule\s+\w+\s*{[^{}]*(?:{[^{}]*}[^{}]*)*[^{}]*(?:/\S*?}[^/]*?/)*)*})")
+
+    def in_bl(self, value):
+        in_bl = False
+        for bl_param in self.parameter_blacklist:
+            if bl_param.lower() == value.lower():
+                in_bl = True
+        return in_bl
+
+    def url_unparse(self, param_type, parsed_url):
+        if param_type == "GETPARAM":
+            querystring = ""
+        else:
+            querystring = parsed_url.query
+        return urlunparse(
+            (
+                parsed_url.scheme,
+                parsed_url.netloc,
+                parsed_url.path,
+                "",
+                querystring if self.retain_querystring else "",
+                "",
+            )
         )
 
+    class ParameterExtractor(ExcavateRule):
```

Comment:
Should have a docstring explaining how it's different from the other excavate rules (i.e. why it's nested here instead of with the others)